### ISSUE TITLE
Start using git the right way & apply best practices

### ISSUE THREAD

**Issue by fiws** [Torndeco/extdb#101]
Hello! I have some suggestions for this repository that I gathered from other git repos (best practices) and my own git experience. 
I put a list on top so you can check what you changed / want to change. Below that there is the explanation why I would change it like that.

I don't know how experienced you are with git, so I tried to explain it relatively easy and low level. Don't take this personally or anything - I just assumed that you are not so familiar because of the way the repo looks :P
Maybe I just don't understand why you made some choices you made. Anyway - I'm just trying to help :)

I hope you will at least adapt some of my proposed changes. Just ask if you have any questions or need more reasons for a change.

Ordered by priority (imo)

* [ ] Don't keep multiple version in `Release`  - use github releases for this
* [ ] Don't wrap the source code in another `EXTDB` / `EXTDB2` folder
  * [ ] actually don't use a dev branch. Use feature branches
* [ ] rename the `dev` or `stable` branch into `master`
* [ ] consider following [semver](http://semver.org/) for versioning.
* [ ] stop creating merge requests for your own branches
* [ ] consider changing the whole file structure

### Situation
why.. just why? Git is version control. It solves this problem for you. You can access any version any time. You even tag the releases so github makes this even simpler. This repo has a ~130MB git directory because of all these binaries in version control.

### Proposed solution
Be consistent about this. Don't put multiple versions with version numbers in there.
Example file structure:
```
dist
├── linux
│   └── @extDB
│       ├── extDB
└── windows
    ├── @extDB
    │   ├── extDB
    ├── tbb.dll
    └── tbbmalloc.dll
```

### or even better
 Don't have a dist/Release directory at all! You can upload binaries using github releases. This keeps your code directory clean and the repo size small. 
![screenshot from 2015-01-14 20 36 21](https://cloud.githubusercontent.com/assets/3409958/5746023/240862f8-9c2d-11e4-8e54-f5e809853fb6.png)

The result looks like this: https://github.com/fiws/extdb/releases/tag/v1.0.1

### Situation
The repository already is named extdb - why should there be another extdb folder.
As for `EXTDB2`: this is super strange. This is what branches are for. Don't keep 2 codebases on the same branch, separating them by folders. Stuff like this should never happen: https://github.com/Torndeco/extdb/commit/5441238bbbb6f0b1cffa516c44fa45991f949e86

### Proposed solution
Just put the contents of extdb in the root of the directory. Use branches for new features. 

#### Example
Lets say I want to implement some mongodb fixes. I then create a branch called `mongodb-fixes` and commit my changes:
```                                                                                      
master❯ git checkout -b mongodb-fixes
Switched to a new branch 'mongodb-fixes'                                                                           
mongodb-fixes❯ commit -am "fix this and that"
mongodb-fixes❯
```
which could look like this:
![screenshot from 2015-01-14 20 53 47](https://cloud.githubusercontent.com/assets/3409958/5746423/c5f4b2ea-9c2f-11e4-82ae-37aca34793cc.png)

Now **if** my fix is done I can locally merge it in master again, and no-one will ever notice that the branch even existed. `git checkout master && git merge mongodb-fixes`

Or you could switch back to master and do some work there:
![screenshot from 2015-01-14 21 37 36](https://cloud.githubusercontent.com/assets/3409958/5747103/abf70db0-9c35-11e4-97e7-2fc8b7e37778.png)

Later you can merge the fixes into master. You actually don't need to ever push that branch unless you want other people to help you with that branch. (Or the branch takes a lot of work and you want to have a safe backup of that branch)
Those short living branches also can be rebased easily and often, for cleaner merging.

The way of managing branches like that is called "Feature Branch Workflow". This technique is pretty popular with git. 
You can read more about it here: https://www.atlassian.com/git/tutorials/comparing-workflows/feature-branch-workflow ( for example - or just google it). 
↑ There are also other workflows explained there if you don't like the feature branch one. I would suggest the centralized workflow otherwise. Most of the others are meant for very big repositories.

### Alternative
If EXTDB2 would differ dramatically from EXTDB(1) - like 0% compatible - you can also create a complete separate repository.

## 3. rename the `dev` or `stable` branch to `master`
I don't get why you have no master branch. It is common in git to have one. The only repos I know that don't have one are doing this because of legacy reasons.
You should rename your "main" branch to master as most people (and git tools) expect this.

## 4. consider following [semver](http://semver.org/) for versioning.
Its not very useful to have versions that are just an incrementing number. I only know which version is newer. I don't know if it breaks my code or if it has any new features.
[Semver](http://semver.org/) solves this by the `MAJOR.MINOR.PATCH` format.
TL;DR:
> MAJOR version when you make incompatible API changes,
>
> MINOR version when you add functionality in a backwards-compatible manner, and
>
> PATCH version when you make backwards-compatible bug fixes.
>Additional labels for pre-release and build metadata are available as extensions to the
> MAJOR.MINOR.PATCH format.

## 5. stop creating merge requests for your own branches
This does not really make any sense if nobody else is involved in that branch. You can merge branches locally.

## 6. consider changing the whole file structure
This repo just needs some cleanup in general. See my suggested file structure here: https://github.com/fiws/extdb
I can create a pull request if you want. I did not because I expect that you might not like *everything* I changed. Also, I did not check if it still compiles or anything. I just changed it so the structure is cleaner.

**Comment by Torndeco** [comment]
Maintaining two different repos for extDB + extDB2 is just way to much work...
Seriously talking 2 different versions of  2 (extDB+extDB2) * 2 (Debug+Normal Version) * 2 (Windows + Linux)
It was just easier to have all info in one single Repo.

Regardless extDB will be getting dropped in the next release anyways whenever i get around to it.
https://github.com/Torndeco/extdb/tree/20
You can see the older structured layout was similiar to what you propose and i will be going back to it in next release.

-----------

Dev Branch is just there if anyone wants to help out or test out some code....
Its literally just my current WIP Code, most of the time its in a broken state.

Since there are times we i might stop working on extDB altogether for weeks at a time or months at a time. It allows me to keep a history of my own changes and figure out what i was doing or broken, especially when it comes to code changes that work on windows + not on linux etc...

It also allows someone to help out with code WIP etc, but its been pretty much a solo project.
I did get some nice help setting up the CMake Build System, using it for Windows Builds.

-----------

Version Number is a incrementing number because
of https://community.bistudio.com/wiki/parseNumber
This way arma sqf code can check you are running a new enough verison of extDB.

Also the normal standard version scheme, makes very little sense for extDB

There are several methods for accessing the Database, which are seperated into their own classes in the code.
I pretty much just call them Protocols

DB_RAW -> DB_RAW_V2 -> DB_RAW_V3
DB_RAW_NO_QUOTES_V2-> DB_RAW_V3
DB_BASIC->DB_BASIC_V2  (dropped)
DB_PROCEDURE -> DB_PROCEDURE_V2
DB_CUSTOM ->  DB_CUSTOM_V2 -> DB_CUSTOM_V3 -> DB_CUSTOM_V4 -> DB_CUSTOM_V5

Realistically Arma Coder only need to care about the Protocol Name.
If it gets changed then there is ABI Breakage.
Also if you try to load a Protocol that isnt supported (anymore), extDB will return an error message to SQF.

Since the Protocols are seperated into there own classes, i can have multiple versions of same Protocol.
This way extDB can output warnings in log files alerting people that X Protocol is deprecated and will get dropped in the future. I like to give atleast a months notice..

Currently there is multiple support for the following
DB_RAW_V2 + DB_RAW_NO_QUOTES_V2 + DB_RAW_V3  (done via alias the calls)
DB_CUSTOM_V3 + DB_CUSTOM_V5 ( different classes in code)

Note:
DB_CUSTOM also have internal version number in config files to highlight minor changes to config file. Normally these are just extra new options you can use.
Most of the time It takes less than 5 minutes to update db_custom file to changes.

-----------

For example 

Altis Life RPG uses DB_RAW_V2 & only uses 5:<Unique ID> to fetch results.  (requires extDB v14+)
Last i heard the dev for Altis Life RPG was on v15, people running servers were using anything from v18->29 etc.

While A3Wasteland uses DB_CUSTOM_V3 (requires extDB v23 or v25+)

It makes zero sense to bump a major version number for ABI Breakage for a single Protocol, that has no effect on other Protocols. 

-----------

The only real thing not handled by version changes is minor changes to 
extdb-conf.ini
Which are just minor changes to extDB i.e different log settings +  new connection options i.e compress / secure auth for MySQL.

-----------

The binary uploads is something i am aware off, it is just something i never had time to look into.

**Comment by Kiina** [comment]
Using binary upload would be a huge improvement. If you are currently searching for an old Version (for testing purpose) you have to crawl through pages of commits. If you really wanna know what changed codewise you have to crawl through pages of commits which most of the time consist of Windows Build, Linux Build, Changelog updates, removing old Builds, readme Update (is it necessary to have the current version in the readme?). And the current structure makes it hard to see if the extdb-conf.ini changed. And if i see right the SQF code is only example code? Maybe things like this should be in the wiki instead of the repo.

**Comment by Torndeco** [comment]
I used to keep a couple of last versions of extDB on github.
But there were afew issues so have them removed, to stop average person from downloading a bugged version of extDB.

Plus most of the time people are grabbing older version with issues since it was compiled with vs2010 and doesn't need VC REDIST 2013.
Since not all GSP have VC Redist 2013 installed by default.

There really shouldn't be any reason to grab an older version.
Older builds have issues like not catching exceptions, causing server crashes or database session leak, that would basicly leak memory and crash server that was up for like 20+hours.
Releases are tagged anyways.
Random crashes when using Large Text DataTypes in MySQL

------------

As for SQF Code

The wiki is confusing enough, its nicer for people to see SQF Code of one way to implement extDB.
Plus it helps people see whats going on + then they look @ wiki and can use test application aswell.
Its reduced the amount of times had to show people in Skype PMs and walk them through it.

---------------

extdb-conf.ini really never changes, in fact the only changes in a year were.
Log settings (really never did anything except disable trace logging in debug version, pointless)
Added mysql compress + mysql auth options
None of the changes broke anything really

---------------

Regardless extDB more than likely wont be changing to much in the near future.

A3Wasteland uses multiple backends, treats MySQL as Key Value Storage really. So wont be using any specfic extDB features like Steam API / RCon (if i finish them)

Altis Life RPG hasn't been updated in months

Arma3 Epoch, ServerFiles are locked down, even if you rewrite them to use extDB. 
You can redist the altered files due to license of A3 Epoch Code.

---------------

So unless there is a arma3 mod/addon in the near future.
More than likely i will be going inactive for awhile after the next 1-2 releases whenever i get around to them.

Will look @ binary releases etc
